File: Solutions/1833-find-the-highest-altitude/find-the-highest-altitude.swift
class Solution {

    func largestAltitude(_ gain: [Int]) -> Int {

        // Step 1: Initialize the current altitude
        // The biker starts at altitude 0
        var currentAltitude = 0

        // Step 2: Initialize the maximum altitude found so far
        // Initially, it's the starting altitude
        var maxAltitude = 0

        // Step 3: Loop through each net gain in the gain array
        for delta in gain {

            // Update the current altitude after this segment
            currentAltitude += delta

            // Update the maximum altitude if current is higher
            if currentAltitude > maxAltitude {
                maxAltitude = currentAltitude
            }
        }

        // Step 4: Return the highest altitude reached
        return maxAltitude
    }
}

File: Solutions/1833-find-the-highest-altitude/find-the-highest-altitude.rs
impl Solution {
    pub fn largest_altitude(gain: Vec<i32>) -> i32 {
        // The biker starts at altitude 0
        let mut current_altitude = 0;

        // Since the starting point is also a valid point,
        // initialize the highest altitude as 0.
        let mut highest_altitude = 0;

        // Go through each altitude gain/loss
        for g in gain {
            // Update the current altitude
            current_altitude += g;

            // If the current altitude is higher than any
            // altitude we've seen before, update the answer.
            if current_altitude > highest_altitude {
                highest_altitude = current_altitude;
            }
        }

        // Return the highest altitude reached during the trip
        highest_altitude
    }
}